.github/workflows/tests.yml

```diff
@@ -149,16 +151,43 @@ jobs:
  
       - name: Install dependencies
         run: |
-          sudo apt install tox
+          sudo apt -yqq install tox
           sudo snap install concierge --classic
+          sudo snap install juju-wait --classic
  
       - name: Prepare Environment for functional tests
         run: |
-          # Use concierge to setup LXD and bootstrap juju container
-          sudo concierge prepare -p machine
+          # Use concierge to setup LXD, microk8s and bootstrap juju controller
+          cat <<EOF >>/tmp/concierge.yaml
+          juju:
+            channel: 3.5/stable
+
+          providers:
+            microk8s:
+              enable: true
+              bootstrap: false
+              addons:
+                - dns
+                - hostpath-storage
+                - metallb
+
+            lxd:
+              enable: true
+              bootstrap: true
+          EOF
+
+          sudo concierge prepare -c /tmp/concierge.yaml
+
           # Create second bridged network needed by the functional tests
           lxc network create second --type=bridge
- 
+
+          # Add microk8s as a kubernetes substrate to the Juju controller
+          sudo microk8s status --wait
+          # The kubernetes API is not always immediately available to use, even if
+          # microk8s status reports ready state. These retries ensure that we don't
+          # fail the test unnecessarily
+          (r=3;while ! juju add-k8s mk8s --controller concierge-lxd ; do ((--r))||exit;sleep 10;done)
```

Comment:
10 second sleeps has the potential of wasting a lot of wall clock time, could we perhaps increase the number of retries and lower the sleep time instead, or is the attempted operation heavy?

Reply:
That's a good point. We could do something like 30 retries with 2s sleep.